homeassistant/components/device_tracker/__init__.py

```diff
@@ -196,86 +208,93 @@ def __init__(self, hass: HomeAssistantType, consider_home: timedelta,
-                self.mac_to_dev[mac] = device
+    def async_see(self, mac: str=None, dev_id: str=None, host_name: str=None,
+                  location_name: str=None, gps: GPSType=None,
+                  gps_accuracy=None, battery: str=None, attributes: dict=None):
+        """Notify the device tracker that you see a device.
 
-            device.seen(host_name, location_name, gps, gps_accuracy, battery,
-                        attributes)
+        This method is a coroutine.
+        """
+        if mac is None and dev_id is None:
+            raise HomeAssistantError('Neither mac or device id passed in')
+        elif mac is not None:
+            mac = str(mac).upper()
+            device = self.mac_to_dev.get(mac)
+            if not device:
+                dev_id = util.slugify(host_name or '') or util.slugify(mac)
+        else:
+            dev_id = cv.slug(str(dev_id).lower())
+            device = self.devices.get(dev_id)
 
+        if device:
+            yield from device.async_seen(host_name, location_name, gps,
+                                         gps_accuracy, battery, attributes)
             if device.track:
-                device.update_ha_state()
-
-            self.hass.bus.fire(EVENT_NEW_DEVICE, {
-                ATTR_ENTITY_ID: device.entity_id,
-                ATTR_HOST_NAME: device.host_name,
-            })
-
-            # During init, we ignore the group
-            if self.group is not None:
-                self.group.update_tracked_entity_ids(
-                    list(self.group.tracking) + [device.entity_id])
-            update_config(self.hass.config.path(YAML_DEVICES), dev_id, device)
+                yield from device.async_update_ha_state()
+            return
 
-    def setup_group(self):
-        """Initialize group for all tracked devices."""
-        run_coroutine_threadsafe(
-            self.async_setup_group(), self.hass.loop).result()
+        # If no device can be found, create it
+        dev_id = util.ensure_unique_string(dev_id, self.devices.keys())
+        device = Device(
+            self.hass, self.consider_home, self.track_new,
+            dev_id, mac, (host_name or dev_id).replace('_', ' '))
+        self.devices[dev_id] = device
+        if mac is not None:
+            self.mac_to_dev[mac] = device
+
+        yield from device.async_seen(host_name, location_name, gps,
+                                     gps_accuracy, battery, attributes)
+
+        if device.track:
+            yield from device.async_update_ha_state()
+
+        self.hass.bus.async_fire(EVENT_NEW_DEVICE, {
+            ATTR_ENTITY_ID: device.entity_id,
+            ATTR_HOST_NAME: device.host_name,
+        })
+
+        # During init, we ignore the group
+        if self.group is not None:
+            yield from self.group.async_update_tracked_entity_ids(
+                list(self.group.tracking) + [device.entity_id])
+
+        # update known_devices.yaml
+        yield from self.hass.run_in_executor(
```

Comment:
We should have a lock inside `update_config` because executor threads can run in parallel.

And if we are going to have a lock, we should not wait for it but instead just `async_add_job` it.